webapp/src/js/bootstrapper/index.js

```diff
@@ -136,47 +138,54 @@
     }
 
     translator.setLocale(userCtx.locale);
-    
+
     var username = userCtx.name;
     var localDbName = getLocalDbName(dbInfo, username);
+
     var localDb = window.PouchDB(localDbName, POUCHDB_OPTIONS.local);
+    var remoteDb = window.PouchDB(dbInfo.remote, POUCHDB_OPTIONS.remote);
+
+    let initialReplicationNeeded;
 
     getDdoc(localDb)
       .then(function() {
-        // ddoc found - bootstrap immediately
-        localDb.close();
-        callback();
+        // ddoc found - no need for initial replication
       })
       .catch(function() {
         // no ddoc found - do replication
-
-        var remoteDb = window.PouchDB(dbInfo.remote, POUCHDB_OPTIONS.remote);
-        initialReplication(localDb, remoteDb)
+        initialReplicationNeeded = true;
+        return initialReplication(localDb, remoteDb)
           .then(function() {
             return getDdoc(localDb).catch(function() {
               throw new Error('Initial replication failed');
             });
-          })
-          .then(function() {
-            // replication complete - bootstrap angular
-            setUiStatus('STARTING_APP');
-          })
-          .catch(function(err) {
-            return err;
-          })
-          .then(function(err) {
-            localDb.close();
-            remoteDb.close();
-            if (err) {
-              if (err.status === 401) {
-                return redirectToLogin(dbInfo, err, callback);
-              }
-
-              setUiError();
-            }
-
-            callback(err);
           });
+      })
+      .then(() => purger(localDb, initialReplicationNeeded)
+        .on('start', () => setUiStatus('PURGE_INIT'))
+        .on('progress', function(progress) {
+          setUiStatus('PURGE_INFO', progress);
+        })
+        .on('optimise', () => setUiStatus('PURGE_AFTER')))
+      .then(function() {
+        // replication complete
+        setUiStatus('STARTING_APP');
+      })
```

Comment:
Ok this may be a blocker... if an admin creates a purge function that doesn't compile then the app will fail to start, right? If the app fails to start then it'll never replicate any settings updates, so fixing the function won't make it to the device, which means the phone is bricked.

Can we handle it more gracefully? eg: log the error and start up anyway and try the purge again next time?

Reply:
So this was intentional on my part, but I have been wrong before! We should have philosophical conversations once we're back from the holidays.

I just added a piece of code I planned to do next year around code execution, as it makes my intentions clearer. The situation is now:
 - If there is a parsing error the app doesn't start
 - If there is a execution error the app logs and error[1] and moves on to the next set

The reasoning behind this is:
 - We want to reduce the amount of incorrect / broken configuration in production in general
 - And here especially, as in this instance it would mean people think they are purging when they are not
 - On the one hand we shouldn't stop people from using the app
 - And on the other hand errors just logged to the console are highly likely to be ignored

So:
 - If there is a bad config we can **always** detect we should hard error so it will definitely be caught in testing. We have this philosophy in other places, such as api startup forcing you to have CouchDB configured correctly before it starts.
 - If there is bad config we can only sometimes detect (eg function execution based on parameters etc) we shouldn't hard error, because we aren't confident it will always happen, and so there is a much greater chance it will exist in production.

One addendum here that goes against my reasoning a little is that a hard error could occur because of JS feature support, which may pass if they use their local Chrome and do not also test against the phone app (which is locked to the right version). It is probably conceivable (unfortunately) that people would think to test locally, but not on the device.

[1] I am considering taking the execution errors and putting them somewhere feedback can find them on startup and send out. Maybe that's too extra though?

Reply:
Hard failures are good but only if the phone can recover when the error is fixed on the server. Unless I've missed something if the function fails to compile then potentially every phone will have to be brought back to base and wiped and everything downloaded again. This failure is _too_ hard.

I'm happy with the app not starting up, but only if it can be fixed remotely without IT support. I can't think of any easy way to do that. Maybe starting the app with a parameter which runs replication but obscures the UI. Then when the settings are changed the dialog will show, the user can refresh, and be up and running again.